apis/compute/v1beta1/gkecluster_types.go

```diff
@@ -0,0 +1,1686 @@
+	// +optional
+	// +immutable
+	Description *string `json:"description,omitempty"`
+
+	// EnableKubernetesAlpha: Kubernetes alpha features are enabled on this
+	// cluster. This includes alpha API groups (e.g. v1alpha1) and features that
+	// may not be production ready in the kubernetes version of the master and
+	// nodes. The cluster has no SLA for uptime and master/node upgrades are
+	// disabled. Alpha enabled clusters are automatically deleted thirty days
+	// after creation.
+	// +optional
+	// +immutable
+	EnableKubernetesAlpha *bool `json:"enableKubernetesAlpha,omitempty"`
+
+	// EnableTpu: Enable the ability to use Cloud TPUs in this cluster.
+	// +optional
+	// +immutable
+	EnableTpu *bool `json:"enableTpu,omitempty"`
+
+	// InitialClusterVersion: The initial Kubernetes version for this
+	// cluster.  Valid versions are those
+	// found in validMasterVersions returned by getServerConfig.  The
+	// version can
+	// be upgraded over time; such upgrades are reflected
+	// in
+	// currentMasterVersion and currentNodeVersion.
+	//
+	// Users may specify either explicit versions offered by
+	// Kubernetes Engine or version aliases, which have the following
+	// behavior:
+	//
+	// - "latest": picks the highest valid Kubernetes version
+	// - "1.X": picks the highest valid patch+gke.N patch in the 1.X
+	// version
+	// - "1.X.Y": picks the highest valid gke.N patch in the 1.X.Y version
+	// - "1.X.Y-gke.N": picks an explicit Kubernetes version
+	// - "","-": picks the default Kubernetes version
+	// +optional
+	// +immutable
+	InitialClusterVersion *string `json:"initialClusterVersion,omitempty"`
+
+	// IPAllocationPolicy: Configuration for cluster IP allocation.
+	// +optional
+	// +immutable
+	IPAllocationPolicy *IPAllocationPolicy `json:"ipAllocationPolicy,omitempty"`
+
+	// LabelFingerprint: The fingerprint of the set of labels for this
+	// cluster.
+	// +optional
+	// +immutable
+	LabelFingerprint *string `json:"labelFingerprint,omitempty"`
+
+	// NOTE(hasheddan): LegacyAbac can only be updated via setLegacyAbac
+	// https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters/setLegacyAbac
+
+	// LegacyAbac: Configuration for the legacy ABAC authorization mode.
+	// +optional
+	LegacyAbac *LegacyAbac `json:"legacyAbac,omitempty"`
+
+	// NOTE(hasheddan): Location is labelled as Output Only by GCP but is required
+	// to create a cluster. It is not included in the actual cluster object
+	// itself, but is instead passed to the create call. If a region is given
+	// the cluster will be Regional, if a zone is given the cluster will be
+	// Zonal.
+
+	// Location: The name of the Google Compute
+	// Engine
+	// [zone](/compute/docs/regions-zones/regions-zones#available)
+	// or
+	// [region](/compute/docs/regions-zones/regions-zones#available) in
+	// which
+	// the cluster resides.
+	// +immutable
+	Location string `json:"location"`
+
+	// Locations: The list of Google Compute
```

Comment:
I'm not sure I fully understand it but I guess GCP differentiates where the _cluster_ resides(`Location string`) and where the _nodes_ reside (`Locations []string`). Since we're going with `NodePool`s to define the nodes and [their locations](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.zones.clusters.nodePools/create), we probably don't need to have (`Locations []string`) here.

Reply:
`Location` is mostly used to distinguish if the `GKECluster` is `Zonal` or `Regional`. The `Location` field is actually used as a path parameter in the `Create` [call](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters/create) and is only `[Ouput Only]` on the `*container.Cluster` object itself. `Locations []string` is part of the `*container.Cluster` and is relevant, especially if the the cluster is `Regional`. I am not certain but I believe you can set the zones in `Locations` and it will prevent `NodePool` creation attached to that cluster outside of those zones. It can also be used for selecting a subset of zones in a certain region to be used. I think it would be a violation of high fidelity to remove.

Reply:
I don't really understand. For `locations` array, the doc says `The list of Google Compute Engine zones in which the cluster's nodes should be located. ` If we use inline or standalone `nodePools`, then each node pool has its own location field anyway. If this `locations` array is for inline deprecated `nodeConfig`, then why it's not deprecated as well?

> I believe you can set the zones in `Locations` and it will prevent NodePool creation attached to that cluster outside of those zones

I couldn't find this information in GCP docs or terraform docs. Anyway, I agree that we should keep it unless we get to know that it was there for deprecated inline nodeConfig.